### PR TITLE
Initial inplace ambient upgrade guide

### PR DESCRIPTION
**Pull request by bleggett**
## Description

Basic ambient upgrade guide to go with the install guide. Fixes https://github.com/istio/istio/issues/48387

- I'm only documenting the in-place Helm upgrade 
  - Tested going from 1.19.6 -> 1.20.2
  - This is the minimum viable upgrade with brief traffic disruption for workload pods so it's worth documenting as a baseline even if we improve/change it - this at least should always _work_ and when in-place upgrade is completed, app pods should still be part of the mesh and still function without requiring a restart. 
  - This includes the previously-agreed caveats around using node cordons and blue-green node pools for production blast radius control - I am _not_ documenting that in detail because how it should work and be orchestrated depends on your specific cloud provider/env, and there's no clean way to simulate it with a baseline of `kind` or similar that I know of, and I don't feel good about replicating cloud provider docs.
  - I did not document revision-based installs for `istiod`/gateways because that's still basically an `in-place` upgrade for the datapath in all the ways that users care about, and whether or not you use revisions to install `istiod`/gateway, it would look the same to your apps either way (upgrading ztunnel creates traffic disruption) - using revisions to upgrade ambient doesn't buy the user much of anything right now, because of ztunnel.
  - In current `master`, upgrading Istio CNI in-place also creates app traffic disruption. The proposed [inpod ambient capture mechanism](https://github.com/istio/istio/pull/48253) will allow the Istio CNI agent to be upgraded in-place without disrupting app traffic. But, as mentioned, ztunnel is the last remaining problem in that regard - it doesn't usefully support revisions and cannot be upgraded without traffic disruption or leveraging cloud-provider node pools.

This will need to be slightly revised when `inpod` hits.

## Reviewers

- [x] Ambient
- [x] Docs
- [x] Installation
- [ ] Networking
- [ ] Performance and Scalability
- [ ] Extensions and Telemetry
- [ ] Security
- [ ] Test and Release
- [x] User Experience
- [ ] Developer Infrastructure
- [x] Localization/Translation

### CODE REVIEW

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (dhawton) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:79]:
```suggestion
The ztunnel DaemonSet is the L4 node-proxy component of ambient.

{{< warning >}}
```

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (dhawton) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:99]: Might be worth referencing https://istio.io/latest/docs/setup/additional-setup/cni/#upgrade, it can be upgraded in any order provided it is within 1 minor version of the control plane. So while we don't say before/after, this could be different for ambient.

File: content/en/docs/ops/ambient/install/helm-installation/index.md
Comment (ericvn) [inline content/en/docs/ops/ambient/install/helm-installation/index.md:72]: I think it makes more sense to have the snippet after the See... and warning.

File: content/en/docs/ops/ambient/upgrade/_index.md
Comment (ericvn) [inline content/en/docs/ops/ambient/upgrade/_index.md:4]: This conflicts with install which is also at 2. I suspect we want this to be after that, so maybe 3 or more. But that will conflict with architecture which is currently 3, so need to bump the architecture page to a higher number as well.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (ericvn) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:11]: I think we should change this (noting that you did copy this from the other page):
```suggestion
[installation with Helm](/docs/ops/ambient/install/helm-installation/) with a previous minor or patch version of Istio.
```

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (ericvn) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:13]: Replace with ambient-alpha-warning template.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (ericvn) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:6]: This says `yes` which generates the snippets, but I don't see the test.

**Review by ericvn — COMMENTED** [review]
Quick pass:

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:99]: I'll copy the wording around the version constraint as-is, ty!

File: content/en/docs/ops/ambient/install/helm-installation/index.md
Comment (bleggett) [reply content/en/docs/ops/ambient/install/helm-installation/index.md:72]: It does, and since I copied this block as-is from `content/en/docs/setup/install/helm/index.md` I'll fix it there as well.

File: content/en/docs/ops/ambient/upgrade/_index.md
Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/_index.md:4]: Right, I noticed this and then forgot to fix it.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:13]: Found another spot that should use this too, will roll that up as well.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (dhawton) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:6]: Just need to build some test.sh scripts that call the snips to test them. https://github.com/istio/istio.io/blob/master/tests/README.md

Comment (kfaseela) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:6]: or you can make it `no` for the time being and I can pick up the tests . Anyways I was thinking of doing the integration tests, so docs test can help build an understanding. let me know

Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:6]: I'll do the basic docs tests since it's mostly copy-paste from the existing ambient install tests, we still need proper integ tests with traffic verification.

Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:6]: Done, PTAL.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (frankbu) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:23]: Two `Prerequisites` sections.
image

If you want to just append two items to the boilerplate list:

```suggestion
{{< boilerplate helm-prereqs >}}
```

You'll also need to indent the helm command in step 4 of the boilerplate or the numbering will be off/

Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:23]: Even with the indent it's not numbering correctly - I'm just going to remove the Helm boilerplates as duplicative from the upgrade guide, since we already say to follow the install guide first and the boilerplates are present there already.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (linsun) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:71]: Does it make sense to upgrade istio-cni here after control plane? or before control plane?  Feels it is part of control plane and we are asking users to update it after dataplane atm.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (linsun) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:71]: should we mention node cordon here to minimize disruption?

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (linsun) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:88]: Seems this is required?

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:67]: It does not matter (or rather it should not matter) as long as the skew is respected as per what @dhawton pointed out.

The ordering logic here is

- Required steps
- Optional steps

It is, strictly speaking, optional to upgrade the CNI, so it goes after the CP (which is not optional) for that reason. 

Would be good to have it both ways in docs tests to catch issues anyway.

Comment (linsun) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:67]: Note we have some existing guideline here - https://istio.io/latest/docs/setup/additional-setup/cni/#operation-details

Comment (linsun) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:67]: iirc, our recommendation is to upgrade istio-cni before or after istiod.

**Review by linsun — COMMENTED** [review]
LG - left some comments around order of istio-cni and whether it should be optional.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:67]: For sidecar it matters, and I don't want to change sidecar ordering reqs.

For ambient, it does not, we have no direct control plane dependency in CNI. So I don't want to convey to people that there is a preferred ordering, because there isn't. Both should work and be supported.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:71]: Yeah I should to reinforce.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (linsun) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:67]: Understood it may not matter but we may not have tested all these permutations in the past and I'm a bit cautious on recommending new orderings with unnecessary benefits to users.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:88]: Not strictly speaking. You can upgrade the data plane and control plane and leave the CNI and everything will work just fine if you respect the skew requirements.

That's why I marked it as optional. You can do this now, later, or wait til your next upgrade, depending, as the text (hopefully) makes clear.

Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:88]: Removed `(Optional)` and put CNI next to ztunnel, and after `istiod` (which is where it will need to sit with `inpod` anyway)

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:67]: For inpod, the recommendation will be to upgrade `istio-cni` when `ztunnel` is upgraded, if it is upgraded at all. Istiod isn't a factor.

**Review by dhawton — COMMENTED** [review]
LGTM. As @ericvn had some comments previous I'll defer final approval to him.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (linsun) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:95]: Should we also mention waypoint here?

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (linsun) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md]: Thanks! Just thought of it is simpler not offering a lot of options.

**Review by linsun — COMMENTED** [review]
LGTM from env maintainer perspective.

Asked a small question about waypoint - may want to clarify it. or can wait till waypoint reaching beta.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:95]: We do need it but it's not present in the ambient install guide yet either, and I was trying to keep those at rough parity right now.

We can do as a followup, @kfaseela might want to do that as well since the integration tests should add/include waypoints in the mix, I'd expect.

**Comment by linsun** [comment]
@kfaseela or @ericvn Daniel is on vacationf or 3 days, can you check if this is ready to be merged?

**Comment by kfaseela** [comment]
Can we have an approval from any of the ambient maintainers?

**Review by linsun — APPROVED** [review]

**Comment by linsun** [comment]
> Can we have an approval from any of the ambient maintainers?

Me... sorry thought I approved it already.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (linsun) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:88]: Correct if you want to apply node cordons, you'll need at least 1 nodes, and possibly add extra nodes during upgrade.

Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:88]: Yep. This is all highly dependent on your K8S setup, and it's a bit crude, but it's what we've settled on for the time being.

One of the goals in writing it down in black and white was to clarify what the actual current process for users is, so it's easier for people to reason about future improvements.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (kfaseela) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:105]: Did you mean, "supported"configuration?

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (kfaseela) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:111]:
```suggestion
## Verify the Installation
```

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (kfaseela) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:113]:
```suggestion
### Verify the workload status
```

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (kfaseela) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:46]:
```suggestion
### Manually upgrade the CRDs and Istio base chart
```

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (kfaseela) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:60]:
```suggestion
### Upgrade the Istio discovery Component
```

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (kfaseela) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:68]:
```suggestion
### Upgrade the ztunnel component
```

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (kfaseela) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:81]:
```suggestion
### Upgrade the CNI Component
```

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (kfaseela) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:95]:
```suggestion
### (Optional) Upgrade the Gateway component
```

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (howardjohn) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:6]:
```suggestion
test: yes
status: Experimental
```

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (howardjohn) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:16]: maybe nitpick, but fwiw in sidecar Istio "in place" has no disruption or downtime either. So we may want to make it a bit more clear this is not just losing the benefits of canary, but is also disruptive?

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (howardjohn) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:62]: We do have the ability to canary with control plane ztunnel uses, no? Do we want to encourage that/document it?

Its really like https://istio.io/latest/docs/setup/additional-setup/gateway/#in-place-upgrade -- its "in place" from data plane, but not from control plane

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (howardjohn) [inline content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:85]: This doesn't seem true with ambient anymore, at least in its current stability? Especially with inpod, etc

Comment (howardjohn) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:85]: inpod specifically since it was just released, not fundamental long term issue

**Review by howardjohn — COMMENTED** [review]
Mostly lgtm, just a few small comments. OK to merge

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:105]: Yes, and it's wrong in the install doc too, ty!

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:17]: Yep good point - in sidecar that's because we literally don't upgrade anything and let the sidecar + CP fall into version skew for $period_of_time so it's a bit hidden.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:63]: If we do, I would prefer to add that later on top of this as an alt flow.

Moving the ztunnel to a different CP endpoint requires a manual update of the ztunnel manifest IIRC, and cannot be accomplished thru Helm without rescheduling or weird workarounds where you "upgrade" the old version of the chart with the old version of the chart and hope the resulting manifest diff doesn't cause ztunnel pods to be recycled.

Also, you have to move the node's ztunnel to the new CP revision, which moves the entire node to the new revision - we currently do not have a way in Helm or otherwise to partially upgrade the ztunnel on some nodes, so calling that "canary" is a bit confusing.

We could probably fix that.

Additionally, the ztunnel update itself is still purely inplace and disruptive, and the baseline guide should/must explain how to upgrade everything, at a minimum.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:86]: Correct, this will need to be revised a bit when inpod is fully merged (e.g. https://github.com/istio/istio/pull/48253 - given that is still pending, `master` still is not using `inpod`)

We can hold this PR open until that merges and do it in one go, or we can merge this now and do that later, and I'll followup - no real pref. on ordering.

Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:86]: But yeah long term with inpod the only real hard compatibility requirement is between ztunnel and the CNI agent, not between the CNI agent and the CP.

How we surface that to people if we want to tolerate skew there is TDB.

File: content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md
Comment (zirain) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:91]: can you point out the detail in the doc? At latest we should highlight it.

Comment (bleggett) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:91]: I mention it at the top of the page here: https://github.com/istio/istio.io/pull/14491/files#diff-1e0e52b46d7e94ba4f52d12e3a86d71d14b68b0ade7c78618ef54d81f23d6d18R19

if that's what you're asking.

As I mentioned in the PR description, I don't want to get into the weeds of how this works operationally in this doc, because it is different for every cloud provider, and not easily demo-able locally with `kind` or similar.

That is an inherent downside of this approach, it's implicitly a punt and very cloud-provider dependent.

Comment (zirain) [reply content/en/docs/ops/ambient/upgrade/helm-upgrade/index.md:91]: ok, it's ok for now.

**Review by kfaseela — APPROVED** [review]

**Pull request merged**
